sw/device/lib/dif/dif_uart.h

```diff
@@ -0,0 +1,249 @@
+                                      (even by default) */
+} dif_uart_config_t;
+
+/**
+ * UART instance state
+ *
+ * UART persistent data that is required by all UART API. |base_address| must
+ * be initialised by the caller, before passing into the UART DIF init routine.
+ */
+typedef struct dif_uart {
+  reg32_t base_addr; /**< UART base address */
+} dif_uart_t;
+
+/**
+ * Initialise an instance of UART
+ *
+ * Initialise UART instance using the configuration data in @p arg2.
+ * Information that must be retained, and is required to program UART must be
+ * stored in @p arg3.
+ * @param base_addr Base address of an instance of UART IP block
+ * @param config UART configuration data
+ * @param uart UART state data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_init(uintptr_t base_addr, const dif_uart_config_t *config,
+                   dif_uart_t *uart);
+
+/**
+ * Configure an instance of UART
+ *
+ * Configure UART using the configuration data in @p arg2. This operation
+ * performs fundamental UART configuration.
+ *
+ * @param uart UART state data
+ * @param config UART configuration data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_configure(const dif_uart_t *uart,
+                        const dif_uart_config_t *config);
+
+/**
+ * Set the RX FIFO watermark
+ *
+ * Set the RX FIFO watermark, is only useful when the corresponding interrupt
+ * is enabled. When the queued RX FIFO number of bytes rises to or above this
+ * level, the RX watermark interrupt is raised.
+ *
+ * @param uart UART state data
+ * @param watermark RX FIFO watermark
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_set_watermark_rx(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark);
+/**
+ * Set the TX FIFO watermark
+ *
+ * Set the TX FIFO watermark, is only useful when the corresponding interrupt
+ * is enabled. When the queued TX FIFO number of bytes falls to or below this
+ * level, the TX watermark interrupt is raised.
+ *
+ * @param uart UART state data
+ * @param watermark TX FIFO watermark
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_set_watermark_tx(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark);
+
+/**
+ * Transmit a single UART character (polled)
+ *
+ * Transmit a single UART character @p arg2. This operation is polled, and
+ * will loop until a character can be written into TX FIFO. Shall not be used
+ * inside an UART ISR.
+ *
+ * @param uart UART state data
+ * @param c Character to be transmitted
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_send_byte_polled(const dif_uart_t *uart, uint8_t c);
+
+/**
+ * Receive a single UART character (polled)
+ *
+ * Receive a single UART character and store it in @p arg2. This operation
```

Comment:
I don't think you should be using Doxygen macros like `@p arg2`. They're hard to read in source code, which is how most of this code is going to be written. Prefer the `|var|` form suggested by the style guide, instead.

Reply:
[Public function (API) documentation](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) suggests doxygen style public function description. If this is wrong then the coding style needs updating.

Reply:
So it does. In practice no one does this, so we might want to re-evaluate that; I think it comes down to "how much do we care about Doxygen being able to parse stuff" (ultimately, I suspect very little, once we're mostly writing Rust).

But, you're right in this case, so you can keep it like you've written for this PR.